File: draft/matlab/theorem2.m
%% theorem2.m
% 2个个体(n=2)、2个话题(m=2) 的两组对比实验
% 例1：无竞争依赖 + 结构平衡 -> 非零共识
% 例2：有竞争依赖(符号相反) -> 收敛到 0

clear; clc; close all;

%% 公共设置
n = 2; m = 2;
W = [0.5 0.5;   % 行随机的人际影响矩阵
     0.5 0.5];

% 初始意见(按话题分组)：y = [y1; y2]，每个 yk 是 n×1（所有人对话题k的意见）
y0 = [ 0.8;  -0.3;   % y1(0): 个体1=0.8, 个体2=-0.3
       0.2;   0.6 ]; % y2(0): 个体1=0.2, 个体2=0.6

T  = 40;    % 迭代步数
tol_clip = true;   % 是否每步裁剪到 [-1,1]

%% ===== 例 1：无竞争依赖 + 平衡 =====
% 两个个体的逻辑矩阵 C1, C2 完全一致，无负边
C1 = [1 0; 0.5 0.5];
C2 = [1 0; 0.5 0.5];

A1 = build_A_from_C_and_W({C1,C2}, W);   % 构造(8)中的块矩阵A
[Y_hist1, y_lim1] = iterate_y(A1, y0, T, tol_clip);

figure('Color','w'); 
subplot(1,2,1); hold on; grid on;
plot(0:T, Y_hist1(1,:), 'LineWidth',1.8);   % 话题1-个体1
plot(0:T, Y_hist1(2,:), 'LineWidth',1.8);   % 话题1-个体2
plot(0:T, Y_hist1(3,:), 'LineWidth',1.8);   % 话题2-个体1
plot(0:T, Y_hist1(4,:), 'LineWidth',1.8);   % 话题2-个体2
title('Non competing dependencies' );
xlabel('t'); ylabel('opinion'); ylim([-1.05 1.05]);
legend('y_1^{(1)}','y_1^{(2)}','y_2^{(1)}','y_2^{(2)}','Location','best');

% 计算每个话题的共识值 α_k（最后一步的均值）
alpha1_case1 = mean(Y_hist1(1:2,end));  % 话题1
alpha2_case1 = mean(Y_hist1(3:4,end));  % 话题2
fprintf('alpha_1 ≈ %.4f, alpha_2 ≈ %.4f\n', alpha1_case1, alpha2_case1);

%% ===== 例 2：存在竞争依赖 =====
% 个体1与个体2在 (话题1 -> 话题2) 的符号相反
C1 = [1 0; +0.5 0.5];
C2 = [1 0; -0.5 0.5];

A2 = build_A_from_C_and_W({C1,C2}, W);
[Y_hist2, y_lim2] = iterate_y(A2, y0, T, tol_clip);

subplot(1,2,2); hold on; grid on;
plot(0:T, Y_hist2(1,:), 'LineWidth',1.8);
plot(0:T, Y_hist2(2,:), 'LineWidth',1.8);
plot(0:T, Y_hist2(3,:), 'LineWidth',1.8);
plot(0:T, Y_hist2(4,:), 'LineWidth',1.8);
title('Competing Interdependencies ');
xlabel('t'); ylabel('opinion'); ylim([-1.05 1.05]);
legend('y_1^{(1)}','y_1^{(2)}','y_2^{(1)}','y_2^{(2)}','Location','best');

alpha1_case2 = mean(Y_hist2(1:2,end));
alpha2_case2 = mean(Y_hist2(3:4,end));
fprintf('alpha_1 ≈ %.4f, alpha_2 ≈ %.4f\n', alpha1_case2, alpha2_case2);

%% ===== 函数：由 {C_i} 与 W 构造 (8) 的 A =====
function A = build_A_from_C_and_W(Ccells, W)
    % Ccells: 1×n 的 cell，每个元素是 m×m 的 C_i
    % W: n×n 行随机非负矩阵
    n = numel(Ccells);
    [m, ~] = size(Ccells{1});
    % 先构造 Γ_{kj} = diag(c_{kj,1},...,c_{kj,n})
    Gamma = cell(m,m);
    for k = 1:m
        for j = 1:m
            diagEntries = zeros(n,1);
            for i = 1:n
                diagEntries(i) = Ccells{i}(k,j);
            end
            Gamma{k,j} = diag(diagEntries);
        end
    end
    % 组装块矩阵 A (m×m 个块，每块是 n×n)
    A = zeros(n*m);
    for k = 1:m
        for j = 1:m
            A( (k-1)*n+1:k*n, (j-1)*n+1:j*n ) = Gamma{k,j} * W;
        end
    end
end

%% ===== 函数：迭代 y(t+1) = A y(t) 并记录轨迹 =====
function [Y_hist, y_lim] = iterate_y(A, y0, T, clip01)
    nm = length(y0);
    Y_hist = zeros(nm, T+1);
    Y_hist(:,1) = y0;
    for t = 1:T
        y_next = A * Y_hist(:,t);
        if clip01
            y_next = max(min(y_next,1),-1); % 限制到[-1,1]
        end
        Y_hist(:,t+1) = y_next;
    end
    y_lim = Y_hist(:,end);
end

File: draft/python/theorem2.py
# torch_theorem2.py
# 2 agents (n=2), 2 topics (m=2)
# Case 1: non-competing dependencies -> nonzero consensus
# Case 2: competing dependencies -> consensus to 0 (for the competing topic)

import torch
import matplotlib.pyplot as plt

torch.set_default_dtype(torch.double)

def build_A_from_C_and_W_torch(C_list, W):
    """
    C_list: list of length n, each item is an (m x m) torch tensor (C_i)
    W: (n x n) nonnegative row-stochastic torch tensor
    return: A in (8), shape (nm x nm)
    """
    n = len(C_list)
    m = C_list[0].shape[0]

    # Γ_{kj} = diag(c_{kj,1}, ..., c_{kj,n})
    Gamma = [[None for _ in range(m)] for _ in range(m)]
    for k in range(m):
        for j in range(m):
            diag_entries = torch.tensor([C_list[i][k, j].item() for i in range(n)], dtype=torch.double)
            Gamma[k][j] = torch.diag(diag_entries)

    # Assemble block matrix A (m x m blocks, each n x n)
    A = torch.zeros(n*m, n*m, dtype=torch.double)
    for k in range(m):
        for j in range(m):
            block = Gamma[k][j] @ W
            A[k*n:(k+1)*n, j*n:(j+1)*n] = block
    return A

def iterate_y_torch(A, y0, T, clip_to_unit=True):
    """y(t+1) = A y(t)"""
    nm = y0.numel()
    Y_hist = torch.zeros(nm, T+1, dtype=torch.double)
    Y_hist[:, 0] = y0
    for t in range(T):
        y_next = A @ Y_hist[:, t]
        if clip_to_unit:
            y_next = torch.clamp(y_next, -1.0, 1.0)
        Y_hist[:, t+1] = y_next
    return Y_hist, Y_hist[:, -1]

def run_case(title, C_list, W, y0, T, ax):
    A = build_A_from_C_and_W_torch(C_list, W)
    Y_hist, y_lim = iterate_y_torch(A, y0, T, clip_to_unit=True)

    # plot: y = [y1(agents=1..n); y2(agents=1..n); ...]
    t = torch.arange(T+1).numpy()
    for r in range(Y_hist.shape[0]):
        ax.plot(t, Y_hist[r, :].numpy(), linewidth=1.8)
    ax.set_title(title)
    ax.set_xlabel('t')
    ax.set_ylabel('opinion')
    ax.set_ylim([-1.05, 1.05])
    ax.grid(True)

    # consensus estimates per topic (mean over agents at final time)
    n = W.shape[0]
    m = int(Y_hist.shape[0] / n)
    alphas = []
    for k in range(m):
        idx = slice(k*n, (k+1)*n)
        alphas.append(Y_hist[idx, -1].mean().item())

    return alphas

if __name__ == "__main__":
    # common settings
    n, m = 2, 2
    W = torch.tensor([[0.5, 0.5],
                      [0.5, 0.5]], dtype=torch.double)   # row-stochastic
    y0 = torch.tensor([0.8, -0.3,   # y1(0): agent1=0.8, agent2=-0.3
                       0.2,  0.6],  # y2(0): agent1=0.2, agent2=0.6
                      dtype=torch.double)
    T = 40

    fig, axes = plt.subplots(1, 2, figsize=(11, 4), constrained_layout=True)

    # ---- Case 1: non-competing dependencies (no negative sign conflict)
    C1 = torch.tensor([[1.0, 0.0],
                       [0.5, 0.5]], dtype=torch.double)
    C2 = torch.tensor([[1.0, 0.0],
                       [0.5, 0.5]], dtype=torch.double)
    alphas1 = run_case('Non-competing dependencies',
                       [C1, C2], W, y0, T, axes[0])
    print(f'Case 1 consensus: alpha_1 ≈ {alphas1[0]:.4f}, alpha_2 ≈ {alphas1[1]:.4f}')

    # ---- Case 2: competing interdependencies (opposite signs on 1->2)
    C1c = torch.tensor([[1.0, 0.0],
                        [0.5, 0.5]], dtype=torch.double)
    C2c = torch.tensor([[1.0, 0.0],
                        [-0.5, 0.5]], dtype=torch.double)
    alphas2 = run_case('Competing interdependencies',
                       [C1c, C2c], W, y0, T, axes[1])
    print(f'Case 2 consensus: alpha_1 ≈ {alphas2[0]:.4f}, alpha_2 ≈ {alphas2[1]:.4f}')

    # legend (same order as MATLAB demo)
    axes[0].legend(['y1^(1)','y1^(2)','y2^(1)','y2^(2)'], loc='best')
    axes[1].legend(['y1^(1)','y1^(2)','y2^(1)','y2^(2)'], loc='best')
    plt.show()
